### ISSUE TITLE
Visual script Array front() back() don't have output node

### ISSUE THREAD

**Issue by netcell** [godotengine/godot#11811]
**Operating system or device, Godot version, GPU Model and driver (if graphics related):**
OSX 10.12.5
Godot 3 built at commit 49afad2b4

**Issue description:**
In Visual Script editor
Array `front()` `back()` don't have output data port.
In `Select Method` these functions are noted return `void` (though as I check in the source code, they return `Variant`?)

Expected behavior:
These nodes should have output data ports for the array element returned.

**Steps to reproduce:**
Put an Array variable or Array constant node on to the editor;
Pull the data port and select `Call`;
Select either `void front() const` or `void back() const`;

![screen shot 2017-10-04 at 12 46 53 am](https://user-images.githubusercontent.com/3048558/31142178-afb22c46-a8a3-11e7-8920-f0cdcb53e283.png)

![screen shot 2017-10-04 at 1 23 35 am](https://user-images.githubusercontent.com/3048558/31142191-b4b56c30-a8a3-11e7-9fb3-b74fd17bd86b.png)

**Comment by J08nY** [comment]
This is an issue in [variant_call.cpp#1595](https://github.com/godotengine/godot/blob/master/core/variant_call.cpp#L1595) not being able to set that any Variant might be returned by the method. right?

**Comment by netcell** [comment]
Sound about right, I am not really familiar with Godot source code though.

**Comment by wafemand** [comment]
Also in documentation written that Arrar.front() don't have return value but in cpp implementation in core it return Variant (like a Array,back()). 
![screenshot_35](https://user-images.githubusercontent.com/28672835/31784096-38b49b7a-b509-11e7-9c4d-735c0a761e71.png)
![image](https://user-images.githubusercontent.com/28672835/31784147-5d799064-b509-11e7-8406-1e37fe458db4.png)
https://github.com/godotengine/godot/blob/6a363ad4704cbfa916b0e10491855221a4680b8e/core/array.cpp#L134

**Comment by wafemand** [comment]
Also i find function Transform.xform which also don't have return value in visual script but must.
![image](https://user-images.githubusercontent.com/28672835/31784340-ff51769a-b509-11e7-9c35-32894995ac3f.png)
https://github.com/godotengine/godot/blob/6a363ad4704cbfa916b0e10491855221a4680b8e/core/math/transform.h#L77

**Comment by neikeq** [comment]
The way `PropertyInfo` differentiates between `void` and `Variant` is with `PROPERTY_USAGE_NIL_IS_VARIANT`. I this the `ADDFUNC*` macros should be updated to receive `MethodInfo` instead, which allows passing a `PropertyInfo` as the return type. Then this line:

https://github.com/godotengine/godot/blob/6a363ad4704cbfa916b0e10491855221a4680b8e/core/variant_call.cpp#L157

would be changed to:

```
funcdata.returns = method.return_val.type != Variant::NIL || method.return_val.usage & PROPERTY_USAGE_NIL_IS_VARIANT;
```

**Comment by wafemand** [comment]
Why this code write in
`#ifdef DEBUG_ENABLED`?
Definition of field `returns` also write in DEBUG_ENABLED.
https://github.com/godotengine/godot/blob/f002e4cb82844083edb8a825160582d00f42bb21/core/variant_call.cpp#L57
If it works only in debug configuration then engine don't know that the function return a value or not (if it return a NIL), isn't it?
This is a bag?

**Comment by akien-mga** [comment]
Fixed by #12382.
